Doc/library/functions.rst

```diff
@@ -371,29 +372,66 @@ are always available.  They are listed here in alphabetical order.
 
 
-.. class:: complex(real=0, imag=0)
-           complex(string)
+.. class:: complex(number=0, /)
+           complex(string, /)
+           complex(real=0, imag=0)
 
-   Return a complex number with the value *real* + *imag*\*1j or convert a string
-   or number to a complex number.  If the first parameter is a string, it will
-   be interpreted as a complex number and the function must be called without a
-   second parameter.  The second parameter can never be a string. Each argument
-   may be any numeric type (including complex).  If *imag* is omitted, it
-   defaults to zero and the constructor serves as a numeric conversion like
-   :class:`int` and :class:`float`.  If both arguments are omitted, returns
-   ``0j``.
+   Convert a single string or number to a complex number, or create a
+   complex number from real and imaginary parts.
 
+   If the argument is a string, it should contain a decimal number, optionally
+   preceded by a sign, and optionally followed by the ``j`` or ``J`` suffix,
+   or a pair of decimal numbers, optionally preceded by a sign, separated by
+   the ``+`` or ``-`` operator, and followed by the ``j`` or ``J`` suffix.
+   A string representing a NaN (not-a-number), or infinity is acceptable
+   in place of a decimal number, like in :func:`float`.
+   The string can optionally be surrounded by whitespaces and the round
+   parenthesis ``(`` and ``)``, which are ignored.
+   The string must not contain whitespace between ``+``, ``-``, the ``j`` or
+   ``J`` suffix, and the decimal number.  For example, ``complex('1+2j')`` is fine, but
+   ``complex('1 + 2j')`` raises :exc:`ValueError`.
+
+   If the argument is a number, the constructor serves as a numeric
+   conversion like :class:`int` and :class:`float`.
    For a general Python object ``x``, ``complex(x)`` delegates to
-   ``x.__complex__()``.  If :meth:`~object.__complex__` is not defined then it falls back
-   to :meth:`~object.__float__`.  If :meth:`!__float__` is not defined then it falls back
+   ``x.__complex__()``.
+   If :meth:`~object.__complex__` is not defined then it falls back
+   to :meth:`~object.__float__`.
+   If :meth:`!__float__` is not defined then it falls back
    to :meth:`~object.__index__`.
 
-   .. note::
-
-      When converting from a string, the string must not contain whitespace
-      around the central ``+`` or ``-`` operator.  For example,
-      ``complex('1+2j')`` is fine, but ``complex('1 + 2j')`` raises
-      :exc:`ValueError`.
+   If two arguments are provided or keyword arguments are used, each argument
+   may be any numeric type (including complex).
+   If both arguments are real numbers, return a complex number with the real
+   component *real* and the imaginary component *imag*.
+   If both arguments are complex numbers, return a complex number with the real
+   component ``real.real-imag.imag`` and the imaginary component
+   ``real.imag+imag.real``.
```

Comment:
Presumably this will mention the deprecation if/when #119620  is merged? Or is the plan to remove the mention of allowing complex altogether?

I'd be tempted to not even document this (mis)feature, on the basis that people who already know about it already know about it, and people who don't know about it shouldn't start using it.

Reply:
The plan was to leave this until the end of the deprecation period.

I would be happy to remove this, but then we will receive another reports about wrong documentation or behavior. The old `real + imag*1j` was more correct for input with finite components (and ignoring corner cases with negative zero).

"Each argument may be any numeric type (including complex)." is already in the current documentation, so it is late to hide this without also changing the behavior.

Reply:
Fair enough.

Reply:
Best we can do is to put the deprecation note as close to this sentence, as possible:)